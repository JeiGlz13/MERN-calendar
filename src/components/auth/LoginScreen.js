import React from 'react';
import { useSelector } from 'react-redux';
import banner from "../../assets/agendaWallpaper.jpg";
import { LoginForm } from '../ui/ui_login/LoginForm';
import { RegisterForm } from '../ui/ui_login/RegisterForm';

export const LoginScreen = () => {
    const ui = useSelector(state => state.ui);

    return (
        <section className="w-full bg-gray-200 flex items-center justify-center text-white font-Poppins">
            <div className='w-11/12 my-8 flex flex-row border-8 border-white rounded-3xl z-50 overflow-hidden shadow-2xl'>
                {
                    (ui.register === true)&&(
                        <RegisterForm/>
                    )    
                }
                <div className="lg:flex w-1/2 hidden bg-gray-500 relative items-center animate__animated animate__fadeIn animate__faster" >
                    <img src={banner} alt='banner'
                    className = "relative opacity-90 z-0 overflow-hidden h-screen" />
                    <div class="absolute bg-black opacity-30 inset-0 z-10"></div>
                    
                    <div className="absolute w-full px-24 z-20">
                        <h1 className="text-5xl font-bold text-left tracking-wide">Agenda personal</h1>
                        <p className="text-3xl my-4">Anota tus recordatorios en esta agenda virtual</p>
                    </div>
                
                </div>
                {
                    (ui.login === true)&&(
                        <LoginForm/>
                    )    
                }
            </div>
        </section>
    )
}
