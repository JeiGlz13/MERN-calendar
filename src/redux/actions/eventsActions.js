import Swal from "sweetalert2";
import { fetchWithToken } from "../../helpers/fetch";
import { getEventsByUser } from "../../helpers/getEventsByUser";
import { prepareEvents } from "../../helpers/prepareEvents";
import { types } from "../types/types";

export const eventStartAddNew = (event) =>{
    return async (dispatch, getState) =>{
        const {uid, name} = getState().auth;

        try {
            const resp = await fetchWithToken('events', event, 'POST');
            const body = await resp.json();

            if (body.ok) {
                event.id = body.evento.id;
                event.user = {
                    _id: uid,
                    name: name
                }
                dispatch(eventAddNew(event))
            }
        } catch (error) {
            console.log(error);
        }
    }
}

const eventAddNew = (event) =>({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = (event) =>({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActive = () =>({
    type: types.eventClearActiveEvent
});

export const eventStartUpdate = (event) =>{
    return async (dispatch) =>{
        try {
            const resp = await fetchWithToken(`events/${event.id}`, event, 'PUT');
            const body = await resp.json();

            if (body.ok) {
                dispatch(eventUpdate(event));
            }else{
                Swal.fire('Error', body.msg ,'error');
            }
        } catch (error) {
            console.log(error);
        }
    }
} 

const eventUpdate= (event) =>({
    type: types.eventUpdate,
    payload: event
})

export const eventStartDelete = () =>{
    return async (dispatch, getState) =>{
        const {id} = getState().events.activeEvent;
        try {
            const resp = await fetchWithToken(`events/${id}`, {}, 'DELETE');
            const body = await resp.json();

            if (body.ok) {
                dispatch(eventDeleted());
                Swal.fire('Success', 'Evento eliminado con exito', 'success');
            }else{
                Swal.fire('Error', body.msg ,'error');
            }
        } catch (error) {
            console.log(error);
        }
    }
} 

const eventDeleted = () =>({type: types.eventDeleted});

export const eventSetStartEnd = (start, end) =>({
    type: types.eventSetStartEnd,
    payload: {
        eventStart: start,
        eventEnd: end
    }
});

export const eventClearEvents = () =>({type: types.eventClear});

export const eventClearStartEnd = () => ({type: types.eventClearStartEnd});

export const eventStartLoading = () =>{
    return async (dispatch, getState) =>{
        try {
                const {uid} = getState().auth;
               const resp = await fetchWithToken('events/');
               const body = await resp.json();

               const myEvents = getEventsByUser(body.eventos, uid);

               const eventos = prepareEvents(myEvents);


               dispatch(eventLoaded(eventos));


        } catch (error) {
            console.log(error)
        }
    }
}

const eventLoaded = (eventos) =>({
    type: types.eventLoaded,
    payload: eventos

})