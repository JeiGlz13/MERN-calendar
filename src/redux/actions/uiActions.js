import { types } from "../types/types";

export const setLoginForm = () =>({
    type: types.setLoginForm,
})

export const setRegisterForm = () =>({
    type: types.setRegisterForm,
})