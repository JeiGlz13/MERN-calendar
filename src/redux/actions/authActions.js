import Swal from "sweetalert2";
import { fetchWithoutToken, fetchWithToken } from "../../helpers/fetch";
import { types } from "../types/types";
import { eventClearActive, eventClearEvents } from "./eventsActions";

export const startLogin = (email, password) =>{
    return async (dispatch) =>{
        const resp = await fetchWithoutToken('auth', {email, password}, 'POST');
        const body = await resp.json();
        if (body.ok) {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            dispatch(login({
                uid: body.uid,
                name: body.name
            }))
        }else{
            Swal.fire('Error', body.msg ,'error');
        }
    }
}

export const startRegister = (name, email, password) =>{
        return async (dispatch) =>{
            const resp = await fetchWithoutToken('auth/new', {name, email, password}, 'POST');
            const body = await resp.json();
            if (body.ok) {
                localStorage.setItem('token', body.token);
                localStorage.setItem('token-init-date', new Date().getTime());
    
                dispatch(login({
                    uid: body.uid,
                    name: body.name
                }))
            }else{
                Swal.fire('Error', body.msg ,'error');
            }
        }
}

export const startChecking = () =>{
    return async (dispatch) =>{
        const resp = await fetchWithToken('auth/renew');
            const body = await resp.json();
            Swal.fire({
                title: 'Iniciando Sesion',
                text: 'Please Wait',
                allowOutsideClick: false,
                didOpen: ()=>{
                    Swal.showLoading();
                }
            });
            if (body.ok) {
                localStorage.setItem('token', body.token);
                localStorage.setItem('token-init-date', new Date().getTime());
    
                dispatch(login({
                    uid: body.uid,
                    name: body.name
                }))
            }else{
                dispatch(checkingFinish());
            }
            Swal.close();
    }
}

const checkingFinish = () =>({type: types.authCheckingFinish})

export const login = (user) => ({
    type: types.authLogin,
    payload: user
})

export const startLogout = () =>{
    return  (dispatch) =>{
        localStorage.clear();
        dispatch(logout());
        dispatch(eventClearEvents());
        dispatch(eventClearActive());
    }
}

export const logout = () =>({
    type: types.authLogout
})