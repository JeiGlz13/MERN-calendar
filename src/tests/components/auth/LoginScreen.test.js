import React from 'react';
import {mount} from 'enzyme';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';

import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startLogin, startRegister } from '../../../redux/actions/authActions';
import Swal from 'sweetalert2';



jest.mock('../../../redux/actions/authActions', () => ({
    startLogin: jest.fn(),
    startRegister: jest.fn()
}));

jest.mock('sweetalert2', () => ({
    fire: jest.fn()
}));

const middleware = [thunk];
const mockStore = configureStore(middleware);


describe('Pruebas en el Login Screen', () => {
    beforeEach(()=>{
        jest.clearAllMocks();
    })

   test('Debe mostrarse correctamente', () => {
        const initState = {
            ui: {
                register: true,
                login: false
            }
        };
        
        const store = mockStore(initState);
        store.dispatch = jest.fn();
        
        const wrapper = mount(
            <Provider store={store}>
                <LoginScreen/>
            </Provider>
        );

       expect(wrapper).toMatchSnapshot();
   })
    
   test('Debe llamar el dispatch del login', () => {
    const initState = {
        ui: {
            register: false,
            login: true
        }
    };
    
    const store = mockStore(initState);
    store.dispatch = jest.fn();
    
    const wrapper = mount(
        <Provider store={store}>
            <LoginScreen/>
        </Provider>
    );

    wrapper.find('input[name="email"]').simulate("change",{
        target: {
            name: "email",
            value: "jeiglz13@gmail.com"
        }
    });

    wrapper.find('input[name="password"]').simulate("change",{
        target: {
            name: "password",
            value: "123456"
        }
    });

    wrapper.find('button').at(1).prop('onClick')({
        preventDefault(){}
    });

    expect(startLogin).toHaveBeenCalledWith('jeiglz13@gmail.com', '123456');
   });

   test('No hay registro si las password son diferentes', () => {

        const initState = {
            ui: {
                register: true,
                login: false
            }
        };
        
        const store = mockStore(initState);
        store.dispatch = jest.fn();
        
        const wrapper = mount(
            <Provider store={store}>
                <LoginScreen/>
            </Provider>
        );

        wrapper.find('input[name="nombre"]').simulate("change",{
            target: {
                name: "nombre",
                value: "Jeisson"
            }
        });

        wrapper.find('input[name="email"]').simulate("change",{
            target: {
                name: "email",
                value: "jeiglz13@gmail.com"
            }
        });

        wrapper.find('input[name="password1"]').simulate("change",{
            target: {
                name: "password1",
                value: "123456"
            }
        });

        wrapper.find('input[name="password2"]').simulate("change",{
            target: {
                name: "password2",
                value: "1234567"
            }
        });

        wrapper.find('button').at(1).prop('onClick')({
            preventDefault(){}
        });

        expect(startRegister).toHaveBeenCalledTimes(0);
        expect(Swal.fire).toHaveBeenCalledWith('Error', 'Las contraseÃ±as no coinciden', 'error');
       
   });
   
   test('Registro si las password son iguales', () => {

        const initState = {
            ui: {
                register: true,
                login: false
            }
        };
        
        const store = mockStore(initState);
        store.dispatch = jest.fn();
        
        const wrapper = mount(
            <Provider store={store}>
                <LoginScreen/>
            </Provider>
        );

        wrapper.find('input[name="nombre"]').simulate("change",{
            target: {
                name: "nombre",
                value: "Jeisson"
            }
        });

        wrapper.find('input[name="email"]').simulate("change",{
            target: {
                name: "email",
                value: "jeiglz13@gmail.com"
            }
        });

        wrapper.find('input[name="password1"]').simulate("change",{
            target: {
                name: "password1",
                value: "123456"
            }
        });

        wrapper.find('input[name="password2"]').simulate("change",{
            target: {
                name: "password2",
                value: "123456"
            }
        });

        wrapper.find('button').at(1).prop('onClick')({
            preventDefault(){}
        });

        expect(startRegister).toHaveBeenCalledWith('Jeisson', 'jeiglz13@gmail.com', '123456');
        expect(Swal.fire).not.toHaveBeenCalled();
    
    });
});
