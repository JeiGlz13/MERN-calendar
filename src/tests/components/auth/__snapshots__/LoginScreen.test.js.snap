// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el Login Screen Debe mostrarse correctamente 1`] = `
<section
  className="w-full bg-gray-200 flex items-center justify-center text-white font-Poppins"
>
  <div
    className="w-11/12 my-8 flex flex-row border-8 border-white rounded-3xl z-50 overflow-hidden shadow-2xl"
  >
    <div
      className="animate__animated animate__fadeInLeft animate__faster lg:w-1/2 w-full flex items-center justify-center text-center md:px-16 px-0 z-0 bg-zinc-800"
    >
      <div
        className="absolute lg:hidden z-10 inset-0 bg-gray-500 bg-no-repeat bg-cover items-center"
      >
        <div
          className="absolute bg-black opacity-60 inset-0 z-0"
        />
      </div>
      <div
        className="w-full py-6 z-20"
      >
        <h1
          className="my-6 text-4xl sm:text-5xl text-white font-Newake"
        >
          MERN Calendar
        </h1>
        <form
          action=""
          className="sm:w-9/12 w-full px-4 lg:px-0 mx-auto"
        >
          <div
            className="pb-2 pt-4"
          >
            <input
              className="block w-full p-4 text-lg rounded-2xl bg-black"
              id="nombre"
              name="nombre"
              onChange={[Function]}
              placeholder="Nombre"
              type="text"
              value=""
            />
          </div>
          <div
            className="pb-2 pt-4"
          >
            <input
              className="block w-full p-4 text-lg rounded-2xl bg-black"
              id="email"
              name="email"
              onChange={[Function]}
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            className="pb-2 pt-4"
          >
            <input
              className="block w-full p-4 text-lg rounded-2xl bg-black"
              id="password1"
              name="password1"
              onChange={[Function]}
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            className="pb-2 pt-4"
          >
            <input
              className="block w-full p-4 text-lg rounded-2xl bg-black"
              id="password2"
              name="password2"
              onChange={[Function]}
              placeholder="Repita su contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            className="pb-2 pt-4"
          >
            <h1
              className="text-gray-400 text-sm"
            >
              ¿Ya tiene una cuenta?
              <button
                className="text-gray-200 text-sm cursor-pointer ml-1"
                onClick={[Function]}
              >
                Inicie Sesión
              </button>
            </h1>
          </div>
          <div
            className="px-4 pb-2 pt-4 flex justify-center"
          >
            <button
              className="uppercase block w-10/12 p-4 text-lg rounded-full bg-indigo-500 hover:bg-indigo-600 focus:outline-none"
              onClick={[Function]}
            >
              Registrarse
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      className="lg:flex w-1/2 hidden bg-gray-500 relative items-center animate__animated animate__fadeIn animate__faster"
    >
      <img
        alt="banner"
        className="relative opacity-90 z-0 overflow-hidden h-screen"
        src="agendaWallpaper.jpg"
      />
      <div
        class="absolute bg-black opacity-30 inset-0 z-10"
      />
      <div
        className="absolute w-full px-24 z-20"
      >
        <h1
          className="text-5xl font-bold text-left tracking-wide"
        >
          Agenda personal
        </h1>
        <p
          className="text-3xl my-4"
        >
          Anota tus recordatorios en esta agenda virtual
        </p>
      </div>
    </div>
  </div>
</section>
`;
